### PR TITLE
release-20.1: backupccl: run FK schema migration during RESTORE AOST

### PR DESCRIPTION
**Pull request by dt**
Part of the 19.x to 20.x upgrade included work on how schema metadata
encoded foreign keys to eventually remove the need for them to be attached
to indexes, and instead be stored on the table itself.

This work included migrating existing schemas to reflect this reorganization
and also included 'upgrading' schema metadata as it was RESTORE'd from a backup
that may have backed up the old representation.

However this on-the-fly upgrade was only applied to some of the copies of the
schema stored in the backup, namely the snapshot of 'latest' table descriptors
that is kept in backupManifest.Descriptors. A parallel optional structure called
DescriptorChanges exists in so-called 'revision_history' backups that has not
just the schema as it existed when backed up, but also prior version of it that
were revised during the backup window. The schema migraiton upgraded the schema
in 'Descriptors' but did not upgrade the prior copies of schema in 'DescriptorChanges'.

This meant that if a RESTORE was performed with an AS OF SYSTEM TIME clause, it
would search for the correct copy of the schema as of that time, then restore it,
and thus would use a copy from DescriptorChanges that had not been migrated, instead
of the copy from Descriptors, which the schema migration had processed.

Release note (bug fix): Fix a bug where 'RESTORE ... AS OF SYSTEM TIME' that used 'skip_missing_foreign_keys' when run on a backup from a 19.x or older version would incorrectly restore schema metadata relating to the foreign key leading to errors when using or inspecting that table after RESTORE.

### CODE REVIEW

**Review by ajwerner — COMMENTED** [review]
I like this one. I would *love* a test with a backup from 19.1 with revision history and some fks and all that jazz but will accept it later. 

> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62493#-:-MWW0XhA4q8wN3brxSyJ:bf3idwi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @pbardea)

___
*[pkg/ccl/backupccl/restore_planning.go, line 559 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62493#-MWW0Vmd34er6DdPGmv7:-MWW0Vmd34er6DdPGmv8:brbo1hf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c72158607012aeaf9823da66d51b259d6d8e04a/pkg/ccl/backupccl/restore_planning.go#L559)):*
> ```Go
> }
> 
> func maybeRewriteSchemaID(
> ```

was this unused?

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62493)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @pbardea)

___
*[pkg/ccl/backupccl/restore_planning.go, line 559 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62493#-MWW0Vmd34er6DdPGmv7:-MWWH1ph5XygHub_5I2h:b-k1hy94) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c72158607012aeaf9823da66d51b259d6d8e04a/pkg/ccl/backupccl/restore_planning.go#L559)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

was this unused?
</blockquote></details>

no it was --  accidentally picked it up in a rebase but compiler caught it

File: pkg/ccl/backupccl/restore_old_versions_test.go [hunk 1]
```diff
@@ -10,6 +10,7 @@ package backupccl_test
 
 import (
 	"context"
+	fmt "fmt"
```
Comment (ajwerner) [inline pkg/ccl/backupccl/restore_old_versions_test.go:13]: 🤔

File: pkg/ccl/backupccl/restore_planning.go [hunk 2]
```diff
@@ -517,6 +517,41 @@ func allocateTableRewrites(
 	return tableRewrites, nil
 }
 
+// maybeUpgradeTableDescsInSliceis updates the passed slice of table descriptors
```
Comment (ajwerner) [inline pkg/ccl/backupccl/restore_planning.go:520]: s/Sliceis/Slice/

**Review by ajwerner — APPROVED** [review]
LGTM

**Comment by dt** [comment]
Added a test. Test uncovered what I think was an _existing_ behavior from the the on-the-fly upgrade, that might have been a regression already? not sure.
It looks like `sqlbase.maybeUpgradeForeignKeyRepOnIndex` checks `skipFKsWithNoMatchingTable` equally for inbound and outbound indexes, and errors out if either are not found by the getter if it is false. But I'm pretty sure originally, pre-FK migration stuff, RESTORE only required using the option to skip missing target *outbound* FKs, where the restored table would look e.g. in `SHOW CREATE`. I believe inbound were automatically skipped if they wasn't being restored, including ig they're not in the backup. But if a 20.1 node attempts such a restore, if the 19.x representation is there, I think RESTORE will instead fail with `pq: descriptor not found` unless one passes `skip_missing_foreign_keys`. I donno if this is a big deal -- it only affects you if you have the old repr and are only restoring the referenced side.

It also isn't new here, so I backed out an initial attempt at a fix -- I initially always passed skip as true when upgrading the slice, but that has a _different_ subtle behavior change when restoring 19.x backups: it silently skips missing FKs before the RESTORE check even gets to them. I donno. Anyway, I've convinced myself this is separate from this issue and I'm not sure I completely understand the fk migration stuff enough to know if this is right as is or not.

**Review by ajwerner — APPROVED** [review]
I'm fine with that behavior. The user will get an error which might be opaque but it's easy enough to deal with. Thanks for doing this.

:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62493#-:-MWZGyeK5kaC-KpUDpXu:beb6uwc)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @pbardea)

**Comment by dt** [comment]
TFTRs!

**Comment by ajwerner** [comment]
This needs to be backported (forwardported?) to all subsequent release branches too.

**Comment by dt** [comment]
Probably, at the very least to 20.2 and 21.1. I’m going to add a topic for bulk weekly on how far back we want to maintain backup compatibility— having RESTORE need keep all the code around to migrate arbitrarily old backups isn’t very appealing.

**Pull request merged**